packages/services/src/boost-bonus.service.ts

```diff
@@ -0,0 +1,700 @@
+          amount,
+          expiresAt,
+          transaction,
+        );
+
+      this.#logger.info(
+        {
+          boostBonusId: boost.id,
+          activeCount: appliedCompetitions.active.length,
+          pendingCount: appliedCompetitions.pending.length,
+        },
+        "Completed adding bonus boost",
+      );
+
+      return {
+        boostBonusId: boost.id,
+        userId: user.id,
+        amount,
+        expiresAt,
+        appliedToCompetitions: appliedCompetitions,
+      };
+    };
+
+    // If transaction provided, use it; otherwise create new transaction
+    if (tx) {
+      return executeWithTx(tx);
+    } else {
+      return this.#db.transaction(executeWithTx);
+    }
+  }
+
+  /**
+   * Applies a bonus boost to all eligible competitions.
+   *
+   * This is a helper method called by `addBoostBonus` to apply the boost
+   * to all competitions that meet the eligibility criteria.
+   *
+   * **Eligibility Criteria:**
+   * - Competition status is "active" OR "pending"
+   * - `competition.boostStartDate < expiresAt` (boost window starts before expiration)
+   * - `competition.boostEndDate > now` (boost window hasn't ended yet)
+   * - Competition must have `boostStartDate` and `boostEndDate` set
+   *
+   * **Idempotency:**
+   * - Uses unique key pattern: `bonusBoost=${boostBonusId}&competition=${competitionId}`
+   * - If same boost is applied twice, second attempt is noop
+   *
+   * **Error Handling:**
+   * - If competition is deleted between query and application, skips that competition (logs warning)
+   * - If competition is missing boost dates, skips that competition (logs warning)
+   * - Continues processing remaining competitions even if one fails
+   *
+   * @param userId - User ID who owns the boost
+   * @param wallet - User's wallet address
+   * @param boostBonusId - ID of the bonus boost to apply
+   * @param amount - Amount of boost to award
+   * @param expiresAt - When the boost expires
+   * @param tx - Transaction to use for database operations
+   * @returns Object containing lists of active and pending competition IDs where boost was applied
+   *
+   * @private
+   */
+  private async applyBoostBonusToEligibleCompetitions(
+    userId: string,
+    wallet: string,
+    boostBonusId: string,
+    amount: bigint,
+    expiresAt: Date,
+    tx: Transaction,
+  ): Promise<{ active: string[]; pending: string[] }> {
+    const now = new Date();
+    const activeCompetitions: string[] = [];
+    const pendingCompetitions: string[] = [];
+
+    // Find active and pending competitions (limited to 1000 for safety)
+    const activeComps = await this.#competitionRepository.findByStatus({
+      status: "active",
+      params: { sort: "createdAt", limit: 1000, offset: 0 },
+    });
+    const pendingComps = await this.#competitionRepository.findByStatus({
+      status: "pending",
+      params: { sort: "createdAt", limit: 1000, offset: 0 },
+    });
+
+    // Warn if we hit the safety limit
```

Comment:
nit: not sure if you are being defensive here, but this if is not needed because you already limited the query above. but it's fine to keep it, I guess

Reply:
yea its defensive. I guess it will be a long time before we git 1k active comps or 1k pending comps at once. i'll remove